### PR TITLE
Prototype pixel snapping icons with GL_LINEAR

### PR DESCRIPTION
**Pull request by kkaefer**
Implements https://github.com/mapbox/mapbox-gl-js/issues/8730.

We're currently switching between pixel-grid aligned icons in a binary way. In a follow-up patch, we could implement a gradual switch by making `u_crisp` a float, and scaling it between 0 and 1 whenever the user interacts with the map.

* [ ] More visual tests
* [ ] Verify that it works correctly at other pixel ratios. I've observed that neighboring icons from the image atlas bleed into the icon at the sides and we may potentially have to scale the 1 px border with the pixel ratio (i.e. 2 pixels) to account for the pixel grid alignment.
* [ ] Verify on a large number of devices. I already tested a few different ones and they all work correctly. However, since this relies on floating point precision and rounding, there's a chance that some devices may behave erratically. I tried to take this into account when designing the calculation by using the projected center point, before the extrusion vectors are applied to get identical results for all four vertices of the icon.
* [ ] Potentially rework the way we use the transformation matrices to reduce the number of calculations we have to do; we currently have 3 4-dimensional projection matrices in the shader and I'm sure we can reduce that a bit. The code currently uses the canvas size to "unproject" from GL space (-1...1) to pixel space so that we can obtain the fractional pixel value to shift the texture coordinates.

### CODE REVIEW

File: src/shaders/symbol_icon.vertex.glsl [hunk 1]
```diff
@@ -29,6 +32,13 @@ varying float v_fade_opacity;
 
 #pragma mapbox: define lowp float opacity
 
+vec2 absfract(vec2 value) {
+    vec2 result = fract(value);
+    if (result.x > 0.5) result.x -= 1.0;
+    if (result.y > 0.5) result.y -= 1.0;
+    return result;
```
Comment (astojilj) [inline src/shaders/symbol_icon.vertex.glsl]: Shorter way to write it `return fract(value) - vec2(greaterThan(color, vec2(0.5)));`. Shouldn't matter, though.

File: src/shaders/symbol_icon.vertex.glsl [hunk 2]
```diff
@@ -84,7 +94,15 @@ void main() {
     vec4 projected_pos = u_label_plane_matrix * vec4(a_projected_pos.xy, 0.0, 1.0);
     gl_Position = u_coord_matrix * vec4(projected_pos.xy / projected_pos.w + rotation_matrix * (a_offset / 32.0 * fontScale), 0.0, 1.0);
 
-    v_tex = a_tex / u_texsize;
+    v_tex = a_tex;
+    if (u_crisp && (segment_angle + symbol_rotation == 0.0)) {
+        // move texture loopup to snap to the pixel grid
```
Comment (astojilj) [inline src/shaders/symbol_icon.vertex.glsl]: loopup/lookup

File: src/shaders/symbol_icon.vertex.glsl [hunk 3]
```diff
@@ -84,7 +94,15 @@ void main() {
...
+        vec4 icon_pos = u_coord_matrix * vec4(projected_pos.xy / projected_pos.w, 0.0, 1.0);
+        v_tex += absfract((icon_pos.xy + 1.0) * u_canvas_size * vec2(1.0, -1.0) / 2.0);
+        v_tex += fract(a_offset / 32.0 * fontScale * u_device_pixel_ratio);
```
Comment (astojilj) [inline src/shaders/symbol_icon.vertex.glsl]: Is there a need to multiply this by u_coord_matrix before fract, as it is in gl_Position calculation?

File: src/shaders/symbol_icon.vertex.glsl (hunk 1 above)
Comment (kkaefer) [reply src/shaders/symbol_icon.vertex.glsl]: I can try, but many OpenGL implementations I tested are _very_ picky about subtle changes like this and perform rounding differently.

**Comment by kkaefer** [comment]
Related issue: https://github.com/mapbox/mapbox-gl-js/issues/8110 (and https://github.com/mapbox/mapbox-gl-js/pull/8118)

**Comment by hrach** [comment]
Is this going to be fixed in release-c? Really looking for seing this fixed :) Thanks.

**Comment by ryanhamley** [comment]
Closing this as stale

**Pull request closed without merging**